### ISSUE TITLE
Special form `Any` should not be compatible with `type[T]`

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#7082]
The following does not produce an error but it should:

```python
from typing import Any

def func[T](x: type[T]) -> None:
    ...

func(Any) # Should produce error
```

**Comment by kyle-query** [comment]
Can you explain why it should produce an error -- is `type[T]` always a constructor?

**Comment by erictraut** [comment]
The `type[T]` form represents an object that is a subclass of `type`. So yes, it's an object that can be called to instantiate objects of type `T`.

`Any` is a special type form intended to be used in type annotations or as a type argument when specializing generics. It is not designed to be used in value expressions at runtime. 

Note that an expression of type `Any` or `type[Any]` can be assigned to a variable or parameter of type `type[T]`. 

```python
def func2(x: Any, y: type[Any]):
    func(x)  # OK
    func(y)  # OK
```

**Comment by kyle-query** [comment]
> It is not designed to be used in value expressions at runtime.

Does this mean the following would be forbidden, as `Any` is used as a value expression below?

```python
from typing import Any, get_type_args

def f(x: Any) -> int:
    return 0

def test(f: Callable[..., int]):
    for name, t in get_type_hints(f).items():
        if t == Any:
            ...
```

**Comment by erictraut** [comment]
That should work fine. You're not attempting to assign the expression `Any` to a variable or parameter declared as `type[T]` in this example.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.349, which I just published.
